### PR TITLE
feat(api): abort signal support for openai, openai-compatible base, zai, kimi-code (round 2)

### LINKED ISSUE #404: [BUG] Stop does not work on OpenAI Compatible API Provider
### Problem (one or two sentences)

Stopping an in-progress question or task does not abort the underlying HTTP request to the provider. The provider continues processing / generating tokens, invisible to the user.

This bug was already reported on Roo code: https://github.com/RooCodeInc/Roo-Code/issues/10779. It still is present in Zoo code 3.56.0.

### Context (who is affected and when)

OpenAI API users. I use llama.cpp on a iGPU as local solution (a propos, dedicated support for llama.cpp would be welcome at some point) which now continues its lengthy process even though I hit the stop button in Zoo.

### Reproduction steps

1. Configure OpenAI API compatible provider, e.g. local llama.cpp setup
2. Give Zoo a prompt
3. Hit the stop button.

### Expected result

Underlying HTTP request is immediatly aborted

### Actual result

Underlying HTTP request continues until the end, consuming precious resources, while of no benefit to the user.

### Variations tried (optional)

_No response_

### App Version

3.56.0

### API Provider (optional)

OpenAI Compatible

### Model Used (optional)

N/A

### Zoo Code Task Links (optional)

_No response_

### Relevant logs or errors (optional)

```shell
If I quit VS Code, the (stopped, but continuing) request is immediately terminated, as is shown in the log of llama.cpp:


1332.16.787.326 W srv          stop: cancel task, id_task = 153
1332.20.921.482 I slot      release: id  0 | task 153 | stop processing: n_tokens = 4096, truncated = 0
1332.20.921.542 I srv  update_slots: all slots are idle
```

### PR DESCRIPTION
**Pull request by easonLiangWorldedtech**
### Related GitHub Issue

Closes #404 — the OpenAI-compatible cancellation objective, for the OpenAI provider family implemented in this PR (`openai`, `base-openai-compatible-provider` and its inheritors, `zai`, `kimi-code`).

This PR covers only the OpenAI-family portion of the broader provider sweep; the remaining scope lands via sibling PRs in this series:

- Part of #616 — openai-codex #1290; openai-native + openai-compatible #1291; anthropic + anthropic-vertex + xai + minimax #1293; unbound + vercel-ai-gateway + zoo-gateway #1295; native-ollama #1299; vscode-lm #1300; gemini + mistral + lite-llm #1303; lm-studio + qwen-code #1309; poe + requesty #1535/#1537/#1538; openrouter #1545 (bedrock already landed on main via #1292).
- Part of #617 — openai-native and openai-codex via #1291 and #1290; bedrock already landed on main via #1292; native-ollama via #1299.
- Part of #618 — vscode-lm via #1300; gemini and mistral via #1303.

### Description

Round 2 of the abort-signal series: wires request-cancellation signals through the OpenAI family of providers.

- **openai.ts**: all five `client.chat.completions.create` sites (createMessage streaming + non-streaming, O3-family streaming + non-streaming, completePrompt) build their request config through `RequestConfigBuilder`, adopted from the start of this PR — the Azure AI Inference `path` option and the abort signal compose in one builder (`setOption("path", ...)` + `setAbortSignal`). Every catch now normalizes abort failures to the Task.ts contract shape (`name === "AbortError"`, message ending in `aborted`) via an abort-aware `handleOpenAIRequestError`, while non-abort errors keep the existing provider-prefix wrap. Both streaming loops (the main `createMessage` path and the O3-family `handleStreamResponse`) gain the loop-defense contract: a top-of-loop break stops processing buffered chunks that arrived after the caller's abort, and a post-loop check surfaces the Task.ts abort contract when the stream would otherwise end normally on an already-aborted signal.
- **base-openai-compatible-provider.ts**: the shared `createMessage` / `createStream` / `completePrompt` path adopted `RequestConfigBuilder` for signal forwarding and gains the exported abort-aware error helper `handleOpenAIRequestError` (reused by zai.ts). Subclasses that do not override these methods (fireworks, sambanova, baseten) inherit the wiring, including the streaming loop's top-of-loop break and post-loop abort check.
- **zai.ts**: audit finding fixed — the GLM thinking path in `createStream` no longer drops `requestOptions`; the thinking path and the glm-5.3 `completePrompt` path forward a merged signal (external signal + `timeoutMs` via `mergeAbortSignalAndTimeout`).
- **kimi-code.ts**: `completePrompt` no longer drops `CompletePromptOptions` — options are forwarded on both the initial call and the 401 OAuth retry. `createMessage` inherits the openai.ts wiring via metadata passthrough.

Design notes:

- `CompletePromptOptions` is not assignable to `ApiHandlerCreateMessageMetadata` (required `taskId`) — gap G7 — so completePrompt paths use `setOption("signal", mergeAbortSignalAndTimeout(...))` instead of `setAbortSignal(metadata)`.
- Gap G5 (zero timeout must mean "no explicit timeout"): `mergeAbortSignalAndTimeout` treats `timeoutMs <= 0` as no timeout internally, so a `timeoutMs: 0` call site passes no signal rather than a timeout that would abort immediately.
- The OpenAI SDK v5 `RequestOptions` type does not satisfy the builder's `RequestConfigOptionsBase` constraint (its `headers`/`signal` shapes differ), so each provider declares a minimal local `OpenAiRequestConfig` shape as the builder generic parameter.
- Each call builds a fresh request-local config (no class-field abort controller), and the per-entry-point `throwIfAborted` guard rejects before any network I/O when the signal is already aborted.

This branch is STACKED on #1288: the foundation commit `e61feb13e` (generic `RequestConfigBuilder`, `mergeAbortSignalAndTimeout`, `mergeAbortSignals`, `throwIfAborted`) rides inside by design.

### Test Procedure

- `pnpm --dir src exec vitest run api/providers/__tests__/openai.spec.ts api/providers/__tests__/base-openai-compatible-provider.spec.ts api/providers/__tests__/zai.spec.ts api/providers/__tests__/kimi-code.spec.ts` — all green. New per-provider "abort signal wiring" suites cover: signal identity at every create site (including Azure path composition), signal + timeout merging, the `timeoutMs: 0` guard, pre-aborted rejection before any request, SDK `APIUserAbortError` and fetch-level `AbortError` normalization to the Task.ts contract shape, and non-abort provider-prefix wrap regression. Deferred-chunk kill tests (2 in openai.spec.ts incl. the O3-family path, 1 in the base spec) prove the loop defense structurally: the second chunk is released only after the abort, so the top-of-loop break is the only thing that prevents the leak (asserted via a `leaked` collection), and `captureError` proves the post-loop check rejects with the contract message rather than a normal stream end.
- 100% changed-line coverage for the four provider files, measured with `vitest run <specs> --coverage` (v8/lcov) and cross-referenced against the `git diff` added lines.
- `pnpm --dir src exec tsc --noEmit` — exit 0.
- `pnpm --dir src exec eslint --prune-suppressions --max-warnings=0 <changed files>` — zero warnings; one stale suppression entry pruned (kimi-code.spec.ts @typescript-eslint/no-explicit-any 1 -> 0, the spec rewrite removed the only as-any cast); no suppression count increased.

### Pre-Submission Checklist

- [x] **Issue Linked**: This PR is linked to an approved GitHub Issue (see "Related GitHub Issue" above).
- [x] **Scope**: My changes are focused on the linked issue (abort-signal wiring for the OpenAI provider family only; the four providers and their specs).
- [x] **Self-Review**: I have performed a thorough self-review of my code.
- [x] **Testing**: New and/or updated tests have been added to cover my changes (if applicable).
- [ ] **Visual Snapshot** (UI changes only): N/A — no UI changes.
- [x] **Documentation Impact**: I have considered if my changes require documentation updates (see "Documentation Updates" section below).
- [x] **Contribution Guidelines**: I have read and agree to the [Contributor Guidelines](/CONTRIBUTING.md).

### Documentation Updates

- [x] No documentation updates are required.
- [ ] Yes, documentation updates are required. (Please describe what needs to be updated or link to a PR in the docs repository.)

### Gate evidence (local, pre-push at 393b516cc)

- **align check** (zdt, unit = this PR's provider set): the 4 streaming-loop ERRORs on this unit's own loops (openai.ts + base-openai-compatible-provider.ts: missing top-of-loop check / missing post-loop check) are cleared by the loop defense above. Residual ERRORs on nanogpt.ts are main-side drift, not this unit: the file is byte-identical to upstream/main (introduced by main #1590 after this branch's fork point) and is outside the unit's scope. The 8 other series units share the same loop-defense gap; series-wide application is a separate maintainer decision.
- **split contract / verify** (zdt, kind=new, design-issue #404): no duplicate symbols, no foreign content. The only violation is budget-hard (1859 a+d vs the 1000 cap) — pre-existing unit scope already accepted through prior review cycles; re-splitting would re-cut already-merged round-1 units.
- **mutation preflight** (local Stryker on the unit delta): 185 valid / 184 killed / 1 timeout / 0 survived / 0 noCoverage. The single timeout is the zai.ts `setOption("signal", ...)` StringLiteral mutant, a known flake class (the mutant drops the signal, so the abort test can only fail through the vitest timeout).

### Additional Notes

Part of the abort-signal series (round 2). Builds on #674, #901, #1008, and #1288. Addresses #404.

### CODE REVIEW

File: src/api/providers/base-openai-compatible-provider.ts [hunk 1]
```diff
@@ -115,7 +143,12 @@ export abstract class BaseOpenAiCompatibleProvider<ModelName extends string>
 		messages: Anthropic.Messages.MessageParam[],
 		metadata?: ApiHandlerCreateMessageMetadata,
 	): ApiStream {
-		const stream = await this.createStream(systemPrompt, messages, metadata)
+		throwIfAborted(metadata?.abortSignal)
+
+		// Per-request abort wiring (RequestConfigBuilder adoption): subclasses inherit
+		// it by receiving the built config as createStream's requestOptions.
+		const requestConfig = new RequestConfigBuilder<OpenAiRequestConfig>().setAbortSignal(metadata).build()
+		const stream = await this.createStream(systemPrompt, messages, metadata, requestConfig)
```
[the remark this answers is not in the document]
Comment (easonLiangWorldedtech) [reply src/api/providers/base-openai-compatible-provider.ts:132]: Valid finding — fixed in 35c95ea19 (fix(api): normalize abort errors raised during stream iteration). The creation-site catches did not cover errors raised by async-iterator consumption, so a mid-stream abort (APIUserAbortError / fetch-level AbortError) leaked as the raw SDK error instead of the Task.ts contract shape. The stream iteration is now wrapped and normalized through the same abort-aware handleOpenAIRequestError at all three sites: the base provider createMessage for-await loop (which also covers the Z.ai inherited thinking path), the openai streaming createMessage for-await loop, and the o3-family yield* this.handleStreamResponse(stream). Non-abort iteration errors keep the existing provider-prefix wrap. Added four regression tests with iterators that reject with APIUserAbortError after the first chunk (base, openai streaming, o3-family streaming, zai thinking path).

File: src/api/providers/base-openai-compatible-provider.ts [hunk 2]
```diff
@@ -23,6 +25,32 @@ type BaseOpenAiCompatibleProviderOptions<ModelName extends string> = ApiHandlerO
 	defaultTemperature?: number
 }
 
+/** Subset of OpenAI.RequestOptions built per request for the abort-signal wiring. */
+type OpenAiRequestConfig = {
+	signal?: AbortSignal
+}
```
[the remark this answers is not in the document]
Comment (easonLiangWorldedtech) [reply src/api/providers/base-openai-compatible-provider.ts:33]: This finding is not valid against the current code: neither openai.ts nor zai.ts imports or references the OpenAiRequestConfig type from base-openai-compatible-provider.ts. Each provider declares its own local OpenAiRequestConfig type (openai.ts adds a path field for the Azure AI Inference composition; zai.ts and the base declare the signal-only shape), and the type is intentionally module-private — it is only used as the RequestConfigBuilder generic parameter inside that file. There is no named import to resolve, so exporting it is unnecessary; keeping it private matches the design note in the PR description ("each provider declares a minimal local OpenAiRequestConfig shape as the builder generic parameter").

**Comment by easonLiangWorldedtech** [comment]
@coderabbitai resume

**Comment by easonLiangWorldedtech** [comment]
@coderabbitai approve

**Comment by easonLiangWorldedtech** [comment]
@coderabbitai approve

**Comment by easonLiangWorldedtech** [comment]
@coderabbitai approve

**Comment by easonLiangWorldedtech** [comment]
@coderabbitai approve

**Comment by easonLiangWorldedtech** [comment]
@coderabbitai approve

**Comment by easonLiangWorldedtech** [comment]
Sync: merged upstream/main (0d937c050) to clear the merge conflict.

One conflicting file: src/api/providers/__tests__/openai.spec.ts — both sides added tests after "should handle streaming responses" (branch-side streaming edge-case tests from this unit; main-side Extra Body tests from #1350). Resolution keeps both sets (6 tests); no semantic changes, openai.ts merged cleanly.

Local gates on the merged head:
- check-types: 11/11 packages pass
- targeted vitest (openai, base-openai-compatible-provider, zai, kimi-code, + suites touched by the main sync): 12 files / 599 tests pass
- eslint --prune-suppressions --max-warnings=0 on all branch-touched src files: 0 warnings, no count increases
- find-missing-translations: clean

Still stacked on #1288 (foundation commit unchanged); standalone unit diff is unchanged by this sync.

**Comment by easonLiangWorldedtech** [comment]
### Out of Scope Changes check — `src/api/index.ts` evidence

The net PR diff contains no change to `src/api/index.ts`. Verified at head `74e406efeb67e63abbe75051b1b5ad8a87574054`:

- `git diff 4c7474d42..74e406efeb67e63abbe75051b1b5ad8a87574054 -- src/api/index.ts` → empty (current base → head)
- `git diff 8d296deef..74e406efeb67e63abbe75051b1b5ad8a87574054 -- src/api/index.ts` → empty (original base → head)
- `git diff 8d296deef..4c7474d42 -- src/api/index.ts` → empty (upstream main has not modified this file between the original base and the current base)
- The PR's Files Changed list (checked via the API at 2026-09-07T04:03:15.201Z; 17 files, see tracking issue snapshot) does not include `src/api/index.ts`

The file is byte-identical across the original base `8d296deef`, the pre-resync branch head `42bcdab06`, the current base `4c7474d42`, and the current head `74e406efeb67e63abbe75051b1b5ad8a87574054`. There is no indentation-only change left to remove or explain; the check's reference describes a stale analysis state from before the branch resync and no longer applies to the current head.

**Comment by easonLiangWorldedtech** [comment]
@coderabbitai review

**Comment by easonLiangWorldedtech** [comment]
@coderabbitai resume

**Comment by easonLiangWorldedtech** [comment]
**CodeRabbit walkthrough of `a3be5ec4f39` still pending — may need an org-member restart**

The 04:22:08Z resync re-armed the review-gate's review-active condition (labels `awaiting-coderabbit` + `coderabbit-review-active` back once CI went green) and CI is green on `a3be5ec4f39` (6/6 workflows, runs created 04:22:12Z). The gate now correctly reads "Required CI passed. Waiting for automated review of the latest commit." — what is missing is the CodeRabbit walkthrough of `a3be5ec4f39` itself.

This branch shows CodeRabbit's auto-review as paused, and the `@coderabbitai review` / `@coderabbitai resume` chat commands are restricted to organization members (fork-owner attempts were rejected at 04:14:43Z / 04:18:50Z). The 03:35Z walkthrough covered the pre-resync head `bccf293b5` (pre-merge checks 4 passed / 3 warning — all 3 addressed in parent `74e406efe`: coverage fix + direct `handleOpenAIRequestError` tests + Kimi retry `timeoutMs` assertions; evidence posted on this PR).

If CodeRabbit does not pick up the queued range for `a3be5ec4f39` on its own, an org member restarting the review (`@coderabbitai`) would unblock the last remaining step for this PR. No push needed — `a3be5ec4f39` should stay the head until the walkthrough lands.

@taltas @JamesRobert20 @navedmerchant @hannesrudolph @edelauna

**Comment by easonLiangWorldedtech** [comment]
**New head `393b516cc` pushed — all CI green; only remaining step is the CodeRabbit walkthrough (paused, needs an org member)**

This push adds the streaming-loop abort defense that the series alignment check flagged for this unit's own loops, plus the mechanical split/mutation gate evidence.

What changed in `393b516cc`:

- **Loop defense** (openai.ts main + O3-family streaming path, base-openai-compatible-provider.ts): top-of-loop break stops processing buffered chunks pulled after the caller's abort; post-loop check surfaces the Task.ts abort contract instead of a silent normal stream end.
- **Kill tests** (3): the second stream chunk is released only after the abort, so the break is the only thing preventing the leak (asserted via a `leaked` collection); `captureError` proves the post-loop check rejects with the contract message. Removing the break or the post-loop check now fails a specific test.

Local gates at this head (pre-push):

- vitest: 4 provider suites, 300 tests pass
- tsc --noEmit: exit 0; eslint --prune-suppressions --max-warnings=0: clean, no suppression count increases
- align check (zdt): the 4 streaming-loop ERRORs on this unit's own loops are cleared. Residual ERRORs on nanogpt.ts are main-side drift (file is byte-identical to upstream/main; introduced by main #1590 after this branch's fork point) — outside this unit's scope. The 8 other series units share the same loop-defense gap; series-wide application is a separate maintainer decision.
- split contract/verify (zdt, kind=new, design-issue #404): no duplicate symbols, no foreign content; only violation is budget-hard (1859 a+d vs 1000) — pre-existing unit scope accepted through prior review cycles
- mutation preflight (local Stryker on the unit delta): 185 valid / 184 killed / 1 timeout / 0 survived / 0 noCoverage (the timeout is the zai.ts `setOption("signal", ...)` StringLiteral mutant — known flake class, the mutant drops the signal so the abort test can only fail via the vitest timeout)

CI on `393b516cc`: all checks green, including mutation-diff (success) and platform-unit-test (ubuntu + windows).

**Blocker — needs an org member:** CodeRabbit's auto-review is paused on this branch (auto-paused after the earlier review churn; the "Reviews paused" note is in the CR status comment), and `@coderabbitai review` / `resume` chat is restricted to organization members in this repo's .coderabbit.yaml (fork-owner attempts were rejected at 09-07 04:14/04:18Z). An org member can unblock by ticking the "▶️ Resume reviews" checkbox in the CodeRabbit status comment (or running `@coderabbitai resume`). Once the CR walkthrough of `393b516cc` posts, the review-gate workflow will move the PR to `awaiting-maintainer` automatically — no further push is needed.
